### PR TITLE
gcylc - toggleable [visualization] style indicator tags.

### PR DESCRIPTION
**Pull request by hjoliver**
Close #54

@cylc/core - this adds small [vis] color indicators to state icons in gcylc (all views), as shown briefly in Exeter, toggleable under the View menu, off by default.

I think this can be helpful if used carefully, e.g. to show family membership at a glance:

foo

Whether or not it would be better to to color the task names (with concerns about visibility) or switch the colors (state vs vis), is probably a matter of opinion.  Those would be relatively easy changes to make on top of this, if necessary.

### CODE REVIEW

**Comment by hjoliver** [comment]
@matthewrmshin - please review/assign.

File: lib/cylc/config.py [hunk 1]
```diff
@@ -651,14 +664,6 @@ def __init__(self, suite, fpath, template_vars=[], template_vars_file=None,
...
                                    "'key1=value1', 'key2=value2', etc.")
 
-        # (Note that we're retaining 'default node attributes' even
-        # though this could now be achieved by styling the root family,
-        # because putting default attributes for root in the suite.rc spec
-        # results in root appearing last in the ordered dict of node
-        # names, so it overrides the styling for lesser groups and
-        # nodes, whereas the reverse is needed - fixing this would
-        # require reordering task_attr in lib/cylc/graphing.py).
-
```
Comment (hjoliver) [inline lib/cylc/config.py]: This comment is no longer valid (tested).

File: lib/cylc/gui/dot_maker.py [hunk 2]
```diff
@@ -135,114 +153,156 @@
...
+        "xxxxx***********",
+        "xxxxx**.......**",
+        "xxxxx**.wwww..**",
+        "xxxxx**.w.....**",
+        "xxxxx**.www...**",
+        "xxxxx**.w.....**",
+        "xxxxx**.w.....**",
+        "xxxxx**.......**",
+        "xxxxx***********",
+        "xxxxx***********"
```
Comment (hjoliver) [inline lib/cylc/gui/dot_maker.py]: This simply adds some blank space left and top, to make room for the new indicator tags.

**Comment by hjoliver** [comment]
A point to note:  This makes task "dot" icons less shareable - before there was just one icon per state, times two for tasks and families.   Now the dot and tree views store (via a separate DotMaker instance owned by each) a composite icon for each individual task ID, regenerated at each update.  This is probably not an efficiency issue as they're very small and generated all at once (not incrementally inside the tree update loop) but in principle we could store fewer unique icons indexed by [is_family, state, vis_border_col, vis_fill_color].   The DotMaker could also be a singleton, or be held by the shared udpater, but that doesn't seem quite appropriate if you're just using the graph view (which only uses the trivial vis indicators, not the full dot icons).

**Comment by matthewrmshin** [comment]
Really like the look.

File: lib/cylc/gui/updater_graph.py [hunk 3]
```diff
@@ -339,6 +369,9 @@ def update_graph(self):
         needs_redraw = current_id != self.prev_graph_id
 
         if needs_redraw:
+            self.vis_png_map = {}
+            self.regenerate_vis_png_map(self.state_summary)
```
Comment (benfitzpatrick) [inline lib/cylc/gui/updater_graph.py]: Do we need to do all the pngs again? Could we do something with the existing `vis_png_map` and `show_vis_tags` information?

File: lib/cylc/gui/updater_tree.py [hunk 4]
```diff
@@ -179,14 +184,18 @@ def update_gui(self):
 
         """
         self.action_required = False
+        if self.state_summary:
```
Comment (benfitzpatrick) [inline lib/cylc/gui/updater_tree.py]: When I start `cylc gui` in this mode, I get:
```
AttributeError: 'TreeUpdater' object has no attribute 'state_summary'
```
coming from this line. Should we set it as empty further up?

**Comment by benfitzpatrick** [comment]
Yep, I really like the look too., and we're packing some really useful information into the displays now. Our global suite looks really good.

I might argue for a bigger/wider colour in the graph view. Even for a `foo.1` node, the dot could be wider without impacting the node size, and it would be easier to see from far away.

I do wonder if we should have two sets of xpms so that the old setting still looks the same, particularly in the status bar.

File: lib/cylc/gui/updater_graph.py (hunk 3 above)
Comment (hjoliver) [inline lib/cylc/gui/updater_graph.py]: I'll take a look.

File: lib/cylc/gui/updater_tree.py (hunk 4 above)
Comment (hjoliver) [inline lib/cylc/gui/updater_tree.py]: Fixed.

**Comment by hjoliver** [comment]
Fixed the tree view init bug.  I tend to agree with Ben's comments, stand by...

**Comment by hjoliver** [comment]
(I'll come back to this after the current batch of more important PRs are done.)

**Comment by hjoliver** [comment]
Branched rebased.

**Comment by hjoliver** [comment]
Branch rebased, conflicts fixed.  I still need to address the feedback...

**Comment by hjoliver** [comment]
Ugh, this is more than 6 months old now.  I'll attempt to finish it off soon...

**Comment by hjoliver** [comment]
(closing this pending time to revisit it - not urgent).

**Pull request closed without merging**